### DISCUSSION TITLE
Keep WSL2 instance with systemd alive

### DISCUSSION THREAD

**Discussion by hrsto** [microsoft/WSL#9245] [General]
I use WSL2 with systemd and run lots of background services. But WSL2 shuts down itself seconds after i exit all terminals. To prevent this from happening, i have to launch a command like `nohup sleep 100d &`. There must be a way to disable that auto shutdown behavior in a non hacky way; %UserProfile%/.wslconfig `vmIdleTimeout=-1` is not working or not doing what its name implies described.

**Comment by TonalidadeHidrica** [comment]
Hi, I am having the same issue.   Are there plans to fix this issue so that we do not need `nohup sleep`-ing hack?

**Comment by Trevorcat** [comment]
Hi, I've met this issue these days, because I construct a new PC.
There is no issue with my old one. I can just open WSL once, exit the terminal window and the Ubuntu can keep running in the background.
There is no gap between those PCs about the configuration. Is any idea about this?

I found a WA for this, using `tmux` in WSL, entering into `tmux` console, and immediately using `control + B` and pressing `D` to detach tmux window but letting it run in the background. then the WSL will keep alive even if I close all WSL windows. But it's really clumsy.

**Comment by daniel-slater** [comment]
Adding this to your .bashrc makes it a bit less clumsy:

<pre><code>
if ! tmux ls > /dev/null 2>&1; then
        tmux new -d > /dev/null 2>&1
fi
</code></pre>

**Reply by daniel-slater** [reply]
But seriously, it's crazy that we can't just disable it in the .wslconfig file.

**Comment by Biswa96** [comment]
You can force WSL2 to keep running with:

```ini
[wsl2]
vmIdleTimeout=-1

```
in %userprofile%/.wslconfig

See https://github.com/microsoft/WSL/issues/10138

**Reply by daniel-slater** [reply]
Nope- that doesn't work. I've had that in my .wslconfig all along. All that does is keep the VM running in the background. It does not prevent your distribution from stopping after the last terminal closes, even with systemd running.

**Comment by geekhal** [comment]
Any progress about this issue? I have met the same problem too. When I don't operate the wsl for some time, the wsl won't react for anymore, I have to shut it down and restart it.

**Comment by klosterhalfen** [comment]
The following works for me
Type into the terminal:

`wsl --exec dbus-launch true`

Even if all terminals are now closed, WSL2 continues to run for me until I explicitly call `wsl --shutdown`. For example, I can still access my Apache server and other services. After a Windows restart I have to enter the command again, but this can be automated.

I had seen the solution here: [https://github.com/microsoft/WSL/issues/10138](https://github.com/microsoft/WSL/issues/10138)


Update: 
Directly with Bash open:

`wsl --cd ~ -- bash -lc "dbus-launch true; exec bash -l"`

**Reply by VirgilMing** [reply]
This is working for my setup (arch on win 10)

**Comment by alirezanet** [comment]
2026 and still no proper solution for this? 🤔 HELLO MICROSOFT 👋

**Reply by yas99en** [reply]
It is better to set both instanceIdleTimeout and vmIdleTimeout.

```
[general]
instanceIdleTimeout=-1

[wsl2]
vmIdleTimeout=-1
```

**Comment by Andarwinux** [comment]
```
[general]
instanceIdleTimeout=-1
```

**Reply by alirezanet** [reply]
It didn't work for me I created two scripts to keep it on:
WSL:
```sh
#!/usr/bin/env bash
set -euo pipefail

if pgrep -af 'dbus-daemon.*--session' >/dev/null 2>&1; then
  echo 'dbus session daemon already running; nothing to do'
  exit 0
fi

exec dbus-launch true
```
and I run it with this powershell script:

```ps
$ErrorActionPreference = 'Stop'
$Distro = 'Ubuntu'
$Script = '~/.local/bin/wsl-ensure-dbus-keepalive.sh'
$Log = 'C:\Users\AliReza\wsl-keepalive.log'

function Run-WSL($Command) {
  $stamp = Get-Date -Format 'yyyy-MM-dd HH:mm:ss'
  "[$stamp] RUN $Command" | Out-File -FilePath $Log -Append -Encoding utf8
  $output = & wsl.exe -d $Distro -- bash -lc $Command 2>&1
  if ($output) {
    $output | Out-File -FilePath $Log -Append -Encoding utf8
  }
  if ($LASTEXITCODE -ne 0) {
    throw "WSL command failed with exit code ${LASTEXITCODE}: $Command"
  }
}

Run-WSL $Script
```
